linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/InternalModuleIDGenerator.scala

```diff
@@ -0,0 +1,157 @@
+    else names.min
+  }
+
+  /** Builds an ID for the class with name [[name]].
+   *
+   *  The result is guaranteed to be:
+   *  - Different from any public module ID.
+   *  - Different for each ClassName.
+   *  - Deterministic.
+   */
+  def forClassName(name: ClassName): ModuleID = {
+    /* Build a module ID that doesn't collide with others.
+     *
+     * We observe:
+     * - Class names are unique, so they never collide with each other.
+     * - Appending a dot ('.') to a class name results in an illegal class name.
+     *
+     * So we append dots until we hit a ModuleID not used by a public module.
+     *
+     * Note that this is stable, because it does not depend on the order we
+     * iterate over nodes.
+     *
+     * To deal with case-insensitive issues, basically we prefix every
+     * uppercase character with a '-', and we prefix every '-' with a '-' to
+     * avoid clashes. However, that is not good enough, since several
+     * uppercase (and titlecase) code points can case-fold to the same
+     * lowercase letter. Therefore, the complete scheme is:
+     *
+     * - ASCII uppercase letters are prefixed with '-'.
+     * - '-' is prefixed with '-'.
+     * - Non-ASCII uppercase letters are prefixed by '-u' followed by the 6
+     *   hexdigits of their codepoint.
+     */
+
+    val builder = new java.lang.StringBuilder
+
+    // First, encode uppercase characters to avoid accidental case-insensitive clashes
+    val originalNameString = name.nameString
+    val originalNameStringLen = originalNameString.length()
+    var i = 0
+    while (i != originalNameStringLen) {
+      val cp = originalNameString.codePointAt(i)
+      if (cp < 0x80) {
+        // ASCII, fast path
+        if (cp == '-' || (cp >= 'A' && cp <= 'Z'))
+          builder.append('-')
+        builder.append(cp.toChar)
+        i += 1
+      } else {
+        // Non-ASCII, slow path
+        if (Character.isUpperCase(cp) || Character.isTitleCase(cp))
+          new java.util.Formatter(builder).format("-u%06x", Integer.valueOf(cp))
+        builder.appendCodePoint(cp)
+        i += Character.charCount(cp)
+      }
+    }
+
+    // Second, avoid colliding with the public module IDs in `avoidSet`
+    var candidateID = builder.toString()
+    while (avoidSet.contains(candidateID)) {
+      builder.append('.')
+      candidateID = builder.toString()
+    }
+    ModuleID(candidateID)
+  }
+
+  /** Creates a digest-based generator whose prefix is not a prefix of any of
+   *  the IDs in [[avoid]] nor of the public module IDs.
+   */
+  def makeFromDigestGenerator(avoid: Iterable[ModuleID]): FromDigestGenerator = {
+    val DashIndex = 8
+    var prefix = "internal-"
+    for (moduleID <- avoid.iterator ++ publicModuleIDs.iterator) {
+      val id = moduleID.id
+      while (id.length() >= prefix.length() &&
+          id.charAt(DashIndex) == '-' && // fast exit
+          id.substring(0, prefix.length()).equalsIgnoreCase(prefix)) {
```

Comment:
I'm concerned that `equalsIgnoreCase` is inconsistent with the logic in `forClassName`. Notably:

- This completely ignores surrogates (works on Chars, not code points)
- It's unclear to me how this handles title case

Reply:
This is OK because we only ever compare against an ASCII string of the form `internal---`. There is no supplementary code point that is equal ignoring case to these ASCII characters. `forClassName` has to deal with all possible code points.